src/net_processing.cpp

```diff
@@ -2954,10 +2961,16 @@ void ProcessMessage(
                 recentRejects->insert(tx.GetWitnessHash());
             }
         } else {
-            if (tx.HasWitness() || state.GetResult() != TxValidationResult::TX_WITNESS_MUTATED) {
-                // Do not use rejection cache for witness transactions or
-                // witness-stripped transactions, as they can have been malleated.
-                // See https://github.com/bitcoin/bitcoin/issues/8279 for details.
+            if (state.GetResult() != TxValidationResult::TX_WITNESS_STRIPPED) {
+                // Do not add txids of witness transactions or witness-stripped
+                // transactions to the filter, as they can have been malleated;
+                // adding such txids to the reject filter would potentially
+                // interfere with relay of valid transactions from peers that
+                // do not support wtxid-based relay. See
+                // https://github.com/bitcoin/bitcoin/issues/8279 for details.
+                // We can remove this restriction (and always add wtxids to
+                // the filter even for witness stripped transactions) once
+                // wtxid-based relay is broadly deployed.
```

Comment:
I think we should be really careful about including `TX_WITNESS_STRIPPED` to the rejection filter. 

In LN, your channel counterparty knows your own version of the commitment transaction as it has to build it to sign it. It means a malicious counterparty would be able to broadcast a witness _stripped_ version of it at any moment and that way get it added to the rejection filters of any p2p peers. If the full-node used by your LN-node is _non-upgraded_ it will announce broadcast of a witness transaction (the local commitment) and it won't be fetched by the lured peers thus effectively censoring your time-sensitive transactions.

Is this understanding of wtxid tx-relay correct ? I don't think it changes anything that your topology is well-established with wtxid-relay peers, the issue is you not being upgraded and subject to intentional reject filters collisions.

Reply:
Can you explain what you think needs to change in the code here?  

I think the documentation is trying to explain a necessary (but not sufficient) condition for changing the code at some unspecified point in the future.  If you want to argue that we should never change this code because other ecosystem software might break, then that's something worthy of discussion, but I don't know that we need to have that discussion in this PR, given that we're not proposing actually doing that now (nor even proposing a timeline for it).

Reply:
Can you reference this comment/issue directly around the code ?

I think that's a bit different than Matt's previous comment [here](https://github.com/bitcoin/bitcoin/pull/18044#discussion_r378586654) where we only considering censorship of a wtixd-relay _requester_, which should be fine if you have at least 1 honest wtxid-relay peer as you pointed.

The scenario I'm describing is I think different, it's censorship of an non-upgraded wtxid-relay initial _sender_ being only connected to reject-witness-stripped wtxid-relay peers. This sender announcing through txid only won't have its transaction fetched and propagated due to an already existent collision in the rejection filter. Such topology configuration sounds realistic for private nodes.

I don't want to argue that we should never change this but when we do it we give a clear warning and time buffer to the rest of the ecosystem to be sure they upgrade. Because by switching p2p policy rules you effectively introduce a critical vulnerability for any software lying on a non-upgraded full-node. LN security model fundamentally reverse the double-spend problem to a private matter, it's the duty of your _individual_ LN-node and full-node to ensure that your transactions propagate well and confirm quickly. 

And I don't think we have guidelines for this today, it sounds more like an illustration of the lack of a set of design docs for a) how we implement BIPs b) what assumptions ecosystem software can do on stability of such implemented rules that you're describing [here](https://github.com/bitcoin/bitcoin/pull/18044#discussion_r400958023).

Reply:
Do you mean just add a link to your github comment, or something else?  

Here's the current comment, which I think captures the issue you're describing (as far as I understand, what you're talking about is fundamentally the same as the issue we flagged years ago when we first deployed segwit, see #8279):
```
                // Do not add txids of witness transactions or witness-stripped
                // witness-stripped transactions, as they can have been malleated.	                // transactions to the filter, as they can have been malleated;
                // See https://github.com/bitcoin/bitcoin/issues/8279 for details.	                // adding such txids to the reject filter would potentially
                // interfere with relay of valid transactions from peers that
                // do not support wtxid-based relay. See
                // https://github.com/bitcoin/bitcoin/issues/8279 for details.
```

But if there's specific different language or a link you'd like to see in here, please just provide alternate text.

Reply:
Right the scenario I'm describing is leveraging the same original issue from #8279, just concerning censorship of a non-upgraded initial sender.

```
               // Note, any on-top-of-base-layer software (eg LN-node) relying on a non-upgraded to wtxid-relay full-node 
              //  would be subject to censor of its time-sensitive transactions. See https://github.com/bitcoin/bitcoin/pull/18044#discussion_r443419034
             //  Removing this restriction should be done considering ecosystem coordination.
```

I think "broadly" is to dim, ideally you want 99% of upgrade among LN-node's full-nodes.

Reply:
I will update the comment to incorporate your suggestion somewhat, including providing a link to the comments as you request, but I have a philosophical difference of opinion about the implications here of ecosystem software that is dependent on particular p2p behavior (specifically with regard to your "ideally you want 99% of upgrade among LN-node's full nodes" comment).

We have anti-CPU DoS reasons for wanting to get rid of the code that tries to determine whether a transaction is witness stripped (as well as generally simplifying the code that handles transactions that fail mempool acceptance).  Requiring that ecosystem software eventually update -- which can be as simple as putting txid-relay nodes behind wtxid-relay border nodes -- in order to be protected against an esoteric attack, with many years of notice, seems completely reasonable to me.

In particular, if we see that the vast majority of full nodes upgrade to wtxid-relay support over the next few years, but a handful of stubborn lightning implementations have not, I think at some point we could say, too bad, find your own workaround for transaction relay if you're worried about this kind of attack.  For these kinds of non-consensus security issues, I think it's reasonable for this project to provide software and solutions (including documenting our behavior and communicating that to our users) and expect that people worried about those kinds of problems will take care of it themselves.  At some point it should be okay for us to have shipped solutions and given people plenty of time to upgrade, and it's not our responsibility if there are people who have chosen not to follow our advice.